### ISSUE TITLE
Reverse urls

### ISSUE THREAD

**Issue by fcracker79** [jrief/django-angular#150]
I am trying to use the 'reverse urls' as specified in the [documentation](http://django-angular.readthedocs.org/en/latest/reverse-urls.html).

According to the documentation, I have to configure django to support the feature, which can be done in two different ways:
1. Add 'djangular.middlewares.DjangularUrlMiddleware' to MIDDLEWARE_CLASSES in your Django
2. Add url(r'^angular/', include('djangular.urls')) to your main urls.py file

In the first case, I receive the error 'ImportError: No module named 'djangular.middleware'', whereas in the second case I receive the error 'No module named 'djangular.urls''.

Maybe I am missing something, but what is strange is that I can not find any DjangularUrlMiddleware in the whole source code. May it be the case that the class name has changed but the documentation has not been modified accordingly?

Thank you for your support.

**Comment by dasf** [comment]
This is only available in the very latest ``0.7.11`` version. You have to update your ``django-angular``, but the latest version from PyPi seems to be broken (#149).
Before this gets fixed you can use ``pip install git+git://github.com/jrief/django-angular.git@0.7.11`` to install the new version.

**Comment by jrief** [comment]
Sorry for that.
Fixed in 0.7.12

**Comment by fcracker79** [comment]
Thank you for your help.
I am loving your grat project!

**Comment by papasax** [comment]
Hi all,

The new angular side reverse function is great. I kind of like the idea of
the url approach for architectural reasons, but whenever i want to do a
post let's say this way :

$http.post(djangoUrl.reverse('myview', {slug:slug}),
$scope.my_slug).success(function(out_data) {
    if (!djangoForm.setErrors($scope.my_form, out_data.errors))
    {
        // on successful post, redirect onto success page
        var successurl = out_data.success_url;
        if (next && successurl){
            $window.location.href = successurl;
        }
    }
}).error(function() {
    swal("server error");
});


I get an error due to @require_GET in djangulat/urls.py

It just means this method cannot be used with form validation. Do we
really need the @require_GET in in djangulat/urls.py?

The middleware approach does not have this issue.

Regards,

Alexandre Desenfant.



2015-03-08 20:34 GMT+01:00 Mirko Bonasorte <notifications@github.com>:

**Comment by jrief** [comment]
@papasax 
sure!

I just removed the @require_GET decorator.

However it then means that you're mixing a querystring with post data, but this shouldn't be a problem with modern browsers. Can you please counter-check. I then will release 0.7.13.

**Comment by papasax** [comment]
Hi Jacob,

I just tried again and there is an other issue : Now, my view is called
with a GET instead of a POST.
The parameter "out_data" just contains my whole html page.

It's not that easy to fix. I don't know if you can redirect a post and if
yes, i don't know it it's a good idea.

Regards,

2015-03-09 14:07 GMT+01:00 Jacob Rief <notifications@github.com>:

**Comment by jrief** [comment]
In the meantime, please proceed with the middleware approach.

A solution is to do the server-side "reversing" using a pure Ajax GET-request. Then use the result and pass it to the real GET/POST/PUT etc. call.
However, since its asynchronous, this result is a promise, rather than a URL. On the other side, it would be much faster, because you do the reversing only once.

I'll try to implement this approach.

**Comment by papasax** [comment]
Hi jacob,

I have an other issue with the middleware approach when I want to make an
api call with query parameters. The reverse django reverse function doesn't
consider query parameters as part of the url, and the angular version does
the same.

Maybe a simple solution would be to use your url approach, but instead of
returning a redirect, just simply return the url. This way you mimic
exactly the reverse function and I can add my query params in javascript. I
agree to say that there would be some network overhead, but in the redirect
strategy it's exactly the same. It's also more "explicit".

What do you think?

2015-03-09 16:00 GMT+01:00 Jacob Rief <notifications@github.com>:

**Comment by dasf** [comment]
Um what exactly do you mean ``.reverse()`` doesn't consider query params as part of url?

**Comment by papasax** [comment]
Have a look there : this question has been described on stackoverflow:
http://stackoverflow.com/questions/25864935/pass-a-query-parameter-with-django-reverse



2015-03-13 16:27 GMT+01:00 Jakob Košir <notifications@github.com>:

**Comment by dasf** [comment]
``` javascript
$http({
    url: djangoUrl.reverse("some_url")
    method: "GET",
    params: {tag: 'food'}
 })
```
Does a ``GET`` request to ``/some/url/?tag=food``

**Comment by papasax** [comment]
This works perfectly.

Thanks.

2015-03-13 16:35 GMT+01:00 Jakob Košir <notifications@github.com>:

**Comment by dasf** [comment]
I guess that's the same problem as in your post in #144 ?
Just to explain what's going wrong:

The new reversing method encodes url name and args as query string, e.g. ``.reverse('api:books')`` -> ``/angular/reverse/?name=api:books``. And if you manually add a query param with ``+?param=123`` you'll end up with ```/angular/reverse/?name=api:books?param=123``.

Notice the second ``?``, which should be ``&``, I guess that's causing some errors.
It worked with previous url resolving approach, ``.reverse()`` returned a url immediately. 

If you pass arguments to ``params`` it knows when append with ``&`` or ``?``, so just use that.

**Comment by papasax** [comment]
Yes it's the same as #144, Your answer is very clear : You should add it to
the documentation.

I posted here as a programing suggestion to the url approach for returning
the url immediately.


2015-03-13 17:04 GMT+01:00 Jakob Košir <notifications@github.com>:

**Comment by jrief** [comment]
@dasf 
I'm still not really happy with the current solution. For instance, in DRF's API browser, the shown URL is the ``/angular/reverse/...`` and not the real URL.
But my other approaches did not work either: "Send a Redirect Response" and "Route Injection". A third approach I'd like to try, is to do a reverse proxying of the whole request. What do you think of it?

**Comment by dasf** [comment]
Well the ``/angular/url/`` part is a bit of a downside, but it's necessary for this approach (and I can't think of a better one).

What do you mean reverse proxying of the request? That's how it works right now. The middleware acts as a reverse proxy, fetches the real response and returns that to client.

**Comment by simondrabble** [comment]
I know this issue is closed, but this seems related - if you'd rather I open a separate issue, let me know.

I'm using djangoUrl.reverse() as described and it works perfectly, however in certain instances I'd like the browser location to reflect the *actual* URI, not the djangoUrl()'ified URI. In other words, when the user clicks on an ```<a>``` tag whose ng-href is dynamically set by djangoUrl.reverse(), the browser shows

    http://someserver.com/angular/reverse/?djng_url_name=foo:bar&djng_url_kwarg_slug=quux-123

instead of the app URI:

    http://someserver.com/foo/bar/quux-123/

The latter is desirable for several reasons:

* From the point-of-view of Cool URIs Don't Change  - consider an app where several points may direct the user to the same page - some via angular, and others with URIs simply encoded by django. Now there are two URIs for the same page. If the user bookmarks the django_angular form, and then django_angular is replaced by some other technology, the bookmarked URI may no longer be valid.

* The django_angular URI exposes information about the nature and implementation of the app that might provide an attacker with ammunition - it is plainly obvious from just the browser's location bar that the server is running django, and how URLs are namespaced.

* The app URI is immediately associated with my app, instead of looking a little like some remote service. This could be important from the viewpoint of a user who may become suspicious that their session is being hijacked.

* It just looks cleaner.

This can be trivially solved by using an HTTP redirect in the middleware instead of calling the view directly (this will also have additional side-benefits - requesting the same URI once if a 301 redirect is used, e.g). However, from what I've seen, and the discussion above, this might not be desirable for some reason. Can you explain the reasons a redirect is not being used? Since the middleware has all the information it needs to construct a redirect, the fix might be as simple as:

    return HttpResponseRedirect(urls.reverse(...))

instead of the current

    new_request = WSGIRequest(request.environ)
    request.__dict__ = new_request.__dict__

**Comment by jkosir** [comment]
http redirects don't work with ``POST`` requests.

**Comment by simondrabble** [comment]
This is true, so the existing solution might be adequate for POSTs (especially in my app, where all POSTs are ajaxy), but for GETs the redirect is desirable.

**Comment by jkosir** [comment]
Well redirects still have all of the disadvantages you listed and also a tiny additional overhead, one extra http response and request.

The initial implementation worked on client side, by loading urlconf into javascript and using regexes to reverse urls there. That "looks" better, but exposing the urlconf is dangerous and also it tends to get big (20-30kb file for average url setup).

This solution is cleaner, creates virtually no overhead, the only downside is that urls look funky.

**Comment by simondrabble** [comment]
I disagree that the existing solution is significantly cleaner enough to warrant funky URLs. A 301 redirect imposes extremely little overhead since the URL will be looked up exactly once, and the redirect will be seamless enough that bookmarking the wrong URL will be impossible, as will discerning app structure from a URL. Additionally, and most importantly for me, funky URLs is something I'd like to avoid. If anyone is interested in my solution it is available in my public repo.

I would certainly avoid any solution that loads the urlconf anywhere outside of the django code itself.

**Comment by jkosir** [comment]
Even with redirects the initial url will still be ``/angular/reverse/something...``. So what's the point of it?
